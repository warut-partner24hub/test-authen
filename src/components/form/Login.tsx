import { useForm, SubmitHandler } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import Inputs from "../input/Inputs";
import { useRouter } from "next/router";
import { toast, ToastContainer } from "react-toastify";
import { signIn } from "next-auth/react";
import "react-toastify/dist/ReactToastify.css"; // Import CSS for toastify

interface ILoginProps {}

const FormSchema = z.object({
  email: z.string().email("Please enter a valid email"),
  password: z
    .string()
    .min(6, "Password must be more than 6 characters")
    .max(16, "Password should not be more than 16 characters"),
});

type FormSchemaType = z.infer<typeof FormSchema>;

const Login: React.FunctionComponent<ILoginProps> = () => {
  const router = useRouter();
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<FormSchemaType>({
    resolver: zodResolver(FormSchema),
  });

  const onSubmit: SubmitHandler<FormSchemaType> = async (data) => {
    try {
      const resSignIn: any = await signIn("credentials", {
        redirect: false,
        email: data.email,
        password: data.password,
      });

      if (resSignIn.error) {
        toast.error(resSignIn.error);
      } else {
        router.push("/");
      }
    } catch (error) {
      toast.error("An unexpected error occurred. Please try again.");
    }
  };

  return (
    <>
      <ToastContainer />
      <div className="flex flex-col justify-center items-center bg-gray-100">
        <div className="bg-white p-8 rounded shadow-md w-full max-w-md">
          <h1 className="text-2xl font-bold mb-6 text-center">Login Form</h1>
          <form className="space-y-6" onSubmit={handleSubmit(onSubmit)}>
            <Inputs
              name="email"
              label="Email"
              type="text"
              placeholder="Email"
              register={register}
              error={errors.email?.message}
              disable={isSubmitting}
            />
            <Inputs
              name="password"
              label="Password"
              type="password"
              placeholder="Password"
              register={register}
              error={errors.password?.message}
              disable={isSubmitting}
            />
            <p className="text-sm">
              Don't have an account?{" "}
              <span
                className="text-blue-500 cursor-pointer"
                onClick={() =>
                  router.push({
                    pathname: router.pathname,
                    query: { tab: "signup" },
                  })
                }
              >
                Sign up
              </span>
            </p>
            <button
              type="submit"
              className="w-full py-2 bg-blue-500 text-white font-semibold rounded hover:bg-blue-600 transition duration-200"
              disabled={isSubmitting}
            >
              {isSubmitting ? "Submitting..." : "Submit"}
            </button>
          </form>
        </div>
      </div>
    </>
  );
};

export default Login;
