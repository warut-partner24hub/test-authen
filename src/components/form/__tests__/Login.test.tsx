import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Login from "../Login";
import { ToastContainer } from "react-toastify";
import { signIn } from "next-auth/react";
import { useRouter } from "next/router";

// Mock next-auth
jest.mock("next-auth/react", () => ({
  signIn: jest.fn(),
}));

// Mock useRouter
jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

describe("Login Form", () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn(),
    });
  });

  test("renders Login form correctly", () => {
    render(
      <>
        <ToastContainer />
        <Login />
      </>
    );

    expect(screen.getByLabelText("Email")).toBeInTheDocument();
    expect(screen.getByLabelText("Password")).toBeInTheDocument();
  });

  test("shows error messages for invalid input", async () => {
    render(
      <>
        <ToastContainer />
        <Login />
      </>
    );

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(
        screen.getByText("Please enter a valid email")
      ).toBeInTheDocument();
      expect(
        screen.getByText("Password must be more than 6 characters")
      ).toBeInTheDocument();
    });
  });

  test("submits the form with valid input", async () => {
    (signIn as jest.Mock).mockResolvedValue({ error: null });

    render(
      <>
        <ToastContainer />
        <Login />
      </>
    );

    fireEvent.change(screen.getByLabelText("Email"), {
      target: { value: "john@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password"), {
      target: { value: "password123" },
    });

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(signIn).toHaveBeenCalledTimes(1);
      expect(signIn).toHaveBeenCalledWith("credentials", {
        redirect: false,
        email: "john@example.com",
        password: "password123",
      });
    });
  });

  test("shows error message for invalid credentials", async () => {
    (signIn as jest.Mock).mockResolvedValue({ error: "Invalid credentials" });

    render(
      <>
        <ToastContainer />
        <Login />
      </>
    );

    fireEvent.change(screen.getByLabelText("Email"), {
      target: { value: "john@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password"), {
      target: { value: "wrongpassword" },
    });

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(signIn).toHaveBeenCalledTimes(1);
    });

    await waitFor(async () => {
      const errMsg = "An unexpected error occurred. Please try again.";
      expect(await screen.findByText(errMsg)).toBeInTheDocument();
    });
  });
});
