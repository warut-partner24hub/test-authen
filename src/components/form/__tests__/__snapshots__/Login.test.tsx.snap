// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col justify-center items-center bg-gray-100 min-h-screen"
  >
    <div
      class="bg-white p-8 rounded shadow-md w-full max-w-md"
    >
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Login Form
      </h1>
      <form
        aria-label="Login form"
        class="space-y-6"
      >
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="email"
          >
            Email
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="email"
              name="email"
              placeholder="Email"
              type="text"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="password"
          >
            Password
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
            />
          </div>
        </div>
        <p
          class="text-sm"
        >
          Don't have an account?
           
          <button
            aria-label="Sign up"
            class="text-blue-500 cursor-pointer underline focus:outline-none"
            type="button"
          >
            Sign up
          </button>
        </p>
        <button
          class="w-full py-2 bg-blue-500 text-white font-semibold rounded hover:bg-blue-600 transition duration-200 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-opacity-50"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
